### DISCUSSION TITLE
error trying to make blended-rectangle from chap. 9 of IOT dev book

### DISCUSSION THREAD

**Discussion by bap293** [Moddable-OpenSource/moddable#693] [Getting started]
I am getting this error when building blended-rectangle from chap. 9 of IOT dev book.  The hello world gui example works fine.

[2/3] Linking CXX executable xs_esp32.elf
FAILED: xs_esp32.elf
cmd.exe /C "cd . && C:\Users\Bill\.espressif\tools\xtensa-esp32-elf\esp-2020r3-8.4.0\xtensa-esp32-elf\bin\xtensa-esp32-elf-g++.exe  -mlongcalls -Wno-frame-address   @CMakeFiles\xs_esp32.elf.rsp  -o xs_esp32.elf  && cd ."
c:/users/bill/.espressif/tools/xtensa-esp32-elf/esp-2020r3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: xs_esp32.a(xsHost.o): in function `modInstrumentationSystemFreeMemory':
C:\Users\Bill\projects\moddable\xs\platforms\esp/xsHost.c:1219: undefined reference to `modInstrumentationAdjust_'

**Comment by phoddie** [comment]
It looks like you may have tried to run the blended-rectangle example using `mcconfig` instead of `mcrun`. That would fail, perhaps in the way you show. 

The process is described in the section "Running Examples" starting on page 12. In short, to run the blended-rectangle example:

1. Build and install the Poco Host 9 (`ch9-poco/host`) using `mcconfig -d -m -p esp32/moddable_two`. (See pages 357-8)
2. Buitd and install the blended-rectangle example (`ch9-poco/blended-rectangle`) using `mcrun -d -m -p esp32/moddable_two` (See page 13)

**Comment by bap293** [comment]
Host applied successfully via mcconfig.

But error on mcrun:

C:\Users\Bill\projects\iot-product-dev-book\ch9-poco\blended-rectangle>mcrun -d -m -p esp32/moddable_two
Installing mod.### fxCommandReceived: remote operation failed with resultCode 65535
NMAKE : fatal error U1077: 'C:\Users\Bill\projects\moddable\build\bin\win\release\serial2xsbug.EXE' : return code '0xffffffff'
Stop.

**Comment by phoddie** [comment]
If I follow the instructions, everything works correctly. I'm using the current Moddable SDK and a Moddable Two.

I am able to duplicate the error you report if I install the wrong host. Is it possible that you didn't install the Chapter 9 host, or perhaps you installed another host over it after? On your PC Chapter 9 host is at `C:\Users\Bill\projects\iot-product-dev-book\ch9-poco\host`. When you install the host, it traces the following in the xsbug console:

<img width="372" alt="image" src="https://user-images.githubusercontent.com/6856458/131193708-c34c9a28-16f1-484d-88e1-d14f0599960d.png">

**Comment by bap293** [comment]
I tried it again, same thing.
Is there a way to verify my setup?

I am tempted to rebuild the whole thing from the  ground up. I've got a decent windows 7 machine (current machine is win 10 - where the setup is now.)

I do get the "device flashed" message after the host install.

In case this matters - here is my process: after installing the host,  I ctrl-c twice to get back to a cmd prompt, then I do the mcrun on the blended rectangle (I do not close xsbug during this process).

Thank you so much for helping - I had a good career in software dev. (now retired) - and I tend not to give up. I've got some ideas for this platform.

**Comment by bap293** [comment]
Rebuilt entire toolchain and SDK. Same issue.
Wondering if we are on different versions of toolchain, sdk or example code.
I am on Esp-idf 4.3 (note: the espressif version, not the one Moddable-OpenSource/esp-idf version); default branch (Public) of Moddable SDK; appears to be only a master branch for the book examples, so I am reasonably certain I am up to date there.

One possible difference is that I am on Win 10.

**Comment by andycarle** [comment]
Windows 10 should be fine; that's where I develop and test the Windows build. 

@bap293: what device are you using? The symptoms you're seeing can sometimes crop up when we don't have the correct baud rate set for the device. But that shouldn't be an issue with, e.g., a Moddable Two.

One other thing that's worth trying is erasing the ESP32's flash storage. Sometimes you can get stuck in an odd state and erasing can fix the problem. You can do that with:

```
python %IDF_PATH%\components\esptool_py\esptool\esptool.py erase_flash
```

**Comment by bap293** [comment]
I erased the flash (thanks for that info. - was wanting to be able to do that).
Same error during mcrun as above.

I am using the moddable-two

I am able to run the moddable examples that only use mcconfig  - I just can't get the ones from the iot dev book that need the mcrun command after loading the host.

One thing to note: I get a missing file error at the end of the flashing o]f the host: Leaving...
Hard resetting via RTS pin...
Done
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) copied.
        1 file(s) copied.
The system cannot find the file specified.
Launching app. Type Ctrl-C twice after debugging app to close serial2xsbug...

i noticed this after I rebuilt the toolchain.

Is it worth rebuilding the tool chain using this: from git: the  esp-idf
Forked from espressif/esp-idf
Espressif IoT Development Framework. Official development framework for ESP32.
